File: c_files/raylib/space_survivors/src/game_manager.c
#include "../include/entities.h"
#include "../include/game_manager.h"
#include "../include/systems.h"
#include "../include/utils.h"

#include <assert.h>
#include <math.h>

void game_manager_init(GameManager *gm)
{
    assert(gm);

    gm->world     = (World){0};
    gm->player_id = create_player(&gm->world);
    gm->camera    = (Camera2D){
           .offset = (Vector2){GetScreenWidth() / 2.0f, GetScreenHeight() / 2.0f},
           .target = (Vector2){0, 0},
           .rotation = 0.0f,
           .zoom     = 1.0f};
}

void game_manager_draw(const GameManager *gm, const Texture2D background,
                       const Texture2D spritesheet)
{
    assert(gm);

    BeginMode2D(gm->camera);

    utils_draw_tiled_background(background);
    system_draw(&gm->world, spritesheet);

    EndMode2D();
}

void game_manager_update(GameManager *gm, const float dt,
                         const Vector2 mouse_pos_world)
{
    assert(gm);

    const Vector2 *player_pos = &gm->world.transform_list[gm->player_id].pos;
    // set player target
    TargetComponent *player_target = &gm->world.target_list[gm->player_id];
    player_target->pos             = &mouse_pos_world;

    system_handle_input(&gm->world);
    system_set_direction(&gm->world);
    system_rotate(&gm->world);
    system_move(&gm->world, dt);
    system_set_bounds(&gm->world);

    // update camera target
    const float half_w = GetScreenWidth() / 2.0f;
    const float half_h = GetScreenHeight() / 2.0f;

    gm->camera.target.x = fmaxf(half_w, fminf(player_pos->x, half_w));
    gm->camera.target.y = fmaxf(half_h, fminf(player_pos->y, half_h));
}

File: python_files/raylib/space_survivors_ecs/game_manager.py
import pyray as p

import systems as S
from ecs import World
from entities import create_enemy, create_player
from settings import ENEMY_SPAWN_RATE, MAP_SIZE, MAX_ENEMIES
from utils import IntervalTimer, utils_draw_tiled_background


class GameManager:
    def __init__(self) -> None:
        self.world: World = World()
        self.player_id: int = create_player(self.world)
        self.enemy_count: int = 0
        self.enemy_timer: IntervalTimer = IntervalTimer(ENEMY_SPAWN_RATE)

        self.camera: p.Camera2D = p.Camera2D()
        self.camera.offset = p.Vector2(
            p.get_screen_width() / 2.0, p.get_screen_height() / 2.0
        )
        self.camera.target = p.Vector2(0, 0)
        self.camera.rotation = 0.0
        self.camera.zoom = 1.0

    def draw(
        self,
        background: p.Texture,
        spritesheet: p.Texture,
        sprite_dict: dict[str, p.Rectangle],
    ) -> None:
        p.begin_mode_2d(self.camera)

        utils_draw_tiled_background(background)
        S.system_draw(self.world, spritesheet, sprite_dict)

        p.end_mode_2d()

    def update(
        self,
        delta_time: float,
        mouse_pos_world: p.Vector2,
        sprite_dict: dict[str, p.Rectangle],
    ) -> None:
        player_pos: p.Vector2 = self.world.transform_list[self.player_id].pos

        # set player target
        self.world.target_list[self.player_id].pos = mouse_pos_world

        # spawn enemy
        if self.enemy_count < MAX_ENEMIES:
            if self.enemy_count <= 0 or self.enemy_timer.update():
                create_enemy(self.world, player_pos)
                self.enemy_count += 1

        S.system_handle_input(self.world)
        S.system_set_direction(self.world)
        S.system_rotate(self.world)
        S.system_move(self.world, delta_time)
        S.system_set_bounds(self.world, sprite_dict)
        S.system_despawn(self.world)

        # update camera target
        half_w: float = p.get_screen_width() / 2.0
        half_h: float = p.get_screen_height() / 2.0

        self.camera.target.x = max(half_w, min(player_pos.x, MAP_SIZE - half_w))
        self.camera.target.y = max(half_h, min(player_pos.y, MAP_SIZE - half_h))
